src/cider/nrepl/middleware/debug.clj

```diff
@@ -1,69 +1,197 @@
+  type is sent in the message as :input-type."
+  [extras type]
+  (let [key (rand-int 10000000)]
+    (->> (assoc extras
+                :status :need-debug-input
+                :key key
+                :input-type type)
+         (response-for @debugger-message)
+         (transport/send (:transport @debugger-message)))
+    (while (not (@debugger-input key))
+      (java.lang.Thread/sleep 100))
+    (let [input (@debugger-input key)]
+      (swap! debugger-input dissoc key)
+      input)))
+
+(def ^:dynamic *locals*
+  "Bound by the `breakpoint` macro to the local &env."
+  {})
 
-(defn breakpoint
+(defn- eval-with-locals
+  "`eval` form wrapped in a let of the *locals* map."
+  [form]
+  (eval
+   `(let ~(vec (mapcat #(list % `(*locals* '~%)) (keys *locals*)))
+      ~form)))
+
+(defn- read-debug-eval-expression
+  "Read and eval an expression from the client.
+  extras is a map to be added to the message, and prompt is added into
+  the :prompt key."
+  [prompt extras]
+  (-> (assoc extras :prompt prompt)
+      (read-debug :expression)
+      eval-with-locals))
+
+(defn- read-debug-command
+  "Read and take action on a debugging command"
+  [value extras]
+  (case (read-debug extras [:next :continue :out :inject :eval :quit])
+    :next     value
+    :continue (do (skip-breaks! true) value)
+    :out      (do (skip-breaks! (butlast (:coor extras))) value)
+    :inject   (read-debug-eval-expression "Expression to inject: " extras)
+    :eval     (let [return (read-debug-eval-expression "Expression to evaluate: " extras)]
+                (read-debug-command value (assoc extras :debug-value (pr-str return))))
+    :quit     (abort!)))
+
+;;; ## High-level functions
+(defmacro breakpoint
   "Send value and coordinates to the client through the debug channel.
   Sends a response to the message stored in debugger-message."
   [value extras]
-  ;; If the user has recently replied (quit) to a `break` statement,
-  ;; then debugger will not ask for input. Therefore, have to avoid
-  ;; sending the following message in these cases.
-  (if (->> (t/now) (t/interval @c/*last-quit-at*) t/in-seconds (< c/*skip-repl-if-last-quit-ago*))
-    ;; TODO: If clj-debugger makes the `read` call in `read-fn`
-    ;; configurable (perhaps with a dynamic variable) we would have to
-    ;; divide our breakpoint into two messages.
-    (do (transport/send
-         (:transport @debugger-message)
-         (response-for @debugger-message
-                       (assoc extras
-                              :debug-value (pr-str value)
-                              :breakfunction nil)))
-        ;; Send the actual break.
-        (binding [*out* (new java.io.StringWriter)]
-          (break value)))
-    ;; Notify cider that a (break) is incoming, along with the value
-    ;; of the instrumented sexp and instructions on how to find it in
-    ;; the code.
-    value))
+  `(binding [*locals* ~(into {} (map (fn [[sym bind]] [`(quote ~sym) (.sym bind)]) &env))]
```

Comment:
Eastwood can't lint this, since it uses tools.analyzer, and `&env` is quite different in tools.analyzer. Check out [this commit](https://github.com/razum2um/clj-debugger/commit/377c34165e108035fcfb169fb17bafbb0aa0e34c) for a workaround.